### PR TITLE
[Merged by Bors] - feat: Port/Combinatorics.Hindman

### PR DESCRIPTION
**Pull request by arienmalec**
port of combinatorics.hindman

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by arienmalec** [comment]
See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/typeclass.20instance.20issues.20in.20!4.232280/near/327855330 for the issues here

**Comment by arienmalec** [comment]
> See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/typeclass.20instance.20issues.20in.20!4.232280/near/327855330 for the issues here
> 
> The fact that this happens around `Semigroup` makes me suspect Lean4 issue 2055
> 
> Or [leanprover/lean4#2074](https://github.com/leanprover/lean4/issues/2074)

Seems like this is a red herring. Down to one issue with `exists_idempotent_ultrafilter_le_FP` where the behavior on the Lean 3 side is hard to understand...

**Comment by arienmalec** [comment]
#

> > See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/typeclass.20instance.20issues.20in.20!4.232280/near/327855330 for the issues here
> > The fact that this happens around `Semigroup` makes me suspect Lean4 issue 2055
> > Or [leanprover/lean4#2074](https://github.com/leanprover/lean4/issues/2074)
> 
> Seems like this is a red herring. Down to one issue with `exists_idempotent_ultrafilter_le_FP` where the behavior on the Lean 3 side is hard to understand...

Fixed, but instance search still very slow where it should be fast.

File: Mathlib/Combinatorics/Hindman.lean [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+    Continuous (· * V) :=
+  TopologicalSpace.IsTopologicalBasis.continuous ultrafilterBasis_is_basis _ <|
+    Set.forall_range_iff.mpr fun s => ultrafilter_isOpen_basic { m : M | ∀ᶠ m' in V, m * m' ∈ s }
+#align ultrafilter.continuous_mul_left Ultrafilter.continuous_mul_left
+#align ultrafilter.continuous_add_left Ultrafilter.continuous_add_left
+
+namespace Hindman
+
+-- porting note: mathport wants these names to be `fS`, `fP`, etc, but this does violence to
+-- mathmatical naming conventions, as does `fs`, `fp`, so we just followed `mathlib` 3 here
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Hindman.lean:96]:
```suggestion
-- mathematical naming conventions, as does `fs`, `fp`, so we just followed `mathlib` 3 here
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**